coreax/approximation.py

```diff
@@ -338,11 +338,75 @@ def approximate(
         :return: Approximation of the kernel matrix inverse
         """
 
+    def _map_approximate(self) -> Callable[[Array, float, Array], Array]:
+        """Define helper function to map approximate over horizontal array stack."""
+        return jax.vmap(self.approximate, in_axes=(0, None, None))
+
+    def approximate_stack(
+        self, kernel_gramians: Array, regularisation_parameter: float, identity: Array
+    ) -> Array:
+        r"""
+        Approximate the regularised inverses of a horizontal stack of kernel matrices.
+
+        :param kernel_gramian: Horizontal Stack of :math:`n \times n` kernel gram
+            matrices
+        :param regularisation_parameter: Regularisation parameter for stable inversion
+            of array, negative values will be converted to positive
+        :param identity: Block identity matrix
+        :return: Approximation of the kernel matrix inverses
+        """
+        return self._map_approximate()(
+            kernel_gramians, regularisation_parameter, identity
+        )
+
+
+class LeastSquareApproximator(RegularisedInverseApproximator):
+    """
+    Approximate inverse of regularised gramian by solving a least-squares problem.
+
+    Note that this approximator does not give time savings and instead acts as a
+    default option useful for comparing other approximators to.
+
+    :param random_key: Key for random number generation
+    :param rcond: Cut-off ratio for small singular values of 'array'. For the purposes
+        of rank determination, singular values are treated as zero if they are smaller
+        than rcond times the largest singular value of 'array'. The default value of
+        None will use the machine precision multiplied by the largest dimension of the
+        array. An alternate value of -1 will use machine precision.
+    """
+
+    def __init__(self, random_key: KeyArrayLike, rcond: Optional[float] = None):
+        """Initialise LeastSquareApproximator class and validate input."""
+        # Initialise parent
+        super().__init__(random_key=random_key)
+        self.rcond = rcond
+
+        # Validate input
+        if self.rcond is not None:
+            if self.rcond < 0 and self.rcond != -1:
+                raise ValueError("'rcond' must be non-negative, except for value of -1")
+
+    @override
+    def approximate(
+        self,
+        kernel_gramian: Array,
+        regularisation_parameter: float,
+        identity: Array,
+    ) -> Array:
+        if kernel_gramian.shape != identity.shape:
+            raise ValueError("Leading dimensions of 'array' and 'identity' must match")
+
+        return jnp.linalg.lstsq(
+            kernel_gramian + abs(regularisation_parameter) * identity,
+            identity,
+            rcond=self.rcond,
+        )[0]
+
 
 def randomised_eigendecomposition(
     random_key: KeyArrayLike,
     array: Array,
-    oversampling_parameter: int = 10,
+    oversampling_parameter: int = 25,
```

Comment:
This default seems to be arbitrary, is there a well-reasoned default that can be computed by making this optional (default None)?

Reply:
From #637:

From the Halko paper, where p is the oversampling_parameter

_Just how many extra samples do we need? Remarkably, for certain types of
random sampling schemes, the failure probability decreases superexponentially with
the oversampling parameter p. As a practical matter, setting p = 5 or p = 10
often gives superb results. This observation is one of the principal facts supporting
the randomized approach to numerical linear algebra._

Also see:

_The number of columns ℓ that the algorithm needs to reach this threshold is usually slightly larger than the minimal rank k of the smallest basis that verifies (4.2; here 4.2. is the error the authors consider for the matrix that has a similar range to the target array). We refer to this discrepancy p = ℓ − k as the oversampling parameter. The size of the oversampling parameter depends on several factors:_

_1. The matrix dimensions. Very large matrices may require more oversampling.
2. The singular spectrum. The more rapid the decay of the singular values, the less oversampling is needed. In the extreme case that the matrix has exact rank k, it is not necessary to oversample._

_Some experimentation led me to use larger values of p for the tests as the kernel matrix I used does not always have a rapidly decaying singular spectrum. The oversampling_parameter  should also not be any larger than the size of the matrix as then you are gaining nothing by this approach.

It could be worth implementing Algorithm 4.2 from the paper as this avoids leaving the choice of oversampling_parameter to the user._

On the power_iterations front Halko has this to say:

_Unfortunately, the simplest version of this scheme (referring to approximate eigen-decomposition) is inadequate in many
applications because the singular spectrum of the input matrix may decay slowly. To
address this difficulty, we incorporate q steps of a power iteration, where q = 1 or
q = 2 usually suffices in practice._

There is an element of diminishing returns, especially with power_iterations, it is one of the more costly parts of the algorithm and once you have reduced the error to machine precision it is pointless to keep beating the horse. I have also found that you can actually make things worse with too many power_iterations, probably due to floating-point errors.

Reply:
Ok that makes sense to me, thanks! It might be useful to point the user to these in the docstrings, or explain the default. On a side note, inside the docstrings, the parameters should be surrounded by backticks instead of single quotes.